docs/context_locals.rst

```diff
@@ -0,0 +1,381 @@
+
+Context locals
+================================================================================
+
+Context local objects ("context locals") are global objects that manage data
+specific to the current greenlet or thread ("the context"). Flask uses
+context locals, like ``request``, ``session``, ``current_app``, and ``g``, for
+data related to the current request ("the request context"). These objects are
+only bound when the application is processing a request.
+
+Motivation
+--------------------------------------------------------------------------------
+
+Many web frameworks, like Django, call view functions with data specific to the
+current request ("the request context"). This is in order to stay thread-safe;
+in these frameworks, view functions generate a response by passing the request
+context throughout their logic.
+
+The problem with this approach is that it can make application logic verbose,
+opaque, and expensive to maintain. Since dependency is transitive, many
+functions will depend on the request context only because they depend on some
+function which depends on the request context. [2]_ This is especially
+frustrating when a dependency on the request context is discovered or eliminated
+in a deeply nested function that causes the signature of all of its callers and
+their callers to be rewritten. This explains why some Django modules fetch the
+request context themselves. [*]_
+
+Many frameworks solve these problems by making the request context globally
+available. However, globals introduce three new problems. First, they can make
+large applications unmaintainable. Second, they aren't thread safe. Third, even
+if we can solve the first two problems, the behavior of thread-safe global is
+complex.
```

Comment:
You're listing these three problems, but not explaining them. e.g. _why_ aren't request-context globals thread safe?

Could you add more to explain these problems? Do you still have questions about them?